language/tools/move-cli/src/main.rs

```diff
@@ -808,30 +808,39 @@ fn doctor(state: OnDiskStateView) -> Result<()> {
         p.parent().unwrap().parent().unwrap().file_name().unwrap()
     }
 
-    let modules = state.get_all_modules()?;
     // verify and link each module
-    for module in modules.values() {
+    let code_cache = state.get_code_cache()?;
+    for module in code_cache.all_modules() {
         if bytecode_verifier::verify_module(module).is_err() {
             bail!("Failed to verify module {:?}", module.self_id())
         }
-        if bytecode_verifier::DependencyChecker::verify_module(module, modules.values()).is_err() {
+
+        let imm_deps = code_cache.get_immediate_module_dependencies(module)?;
+        if bytecode_verifier::DependencyChecker::verify_module(module, imm_deps).is_err() {
             bail!(
                 "Failed to link module {:?} against its dependencies",
                 module.self_id()
             )
         }
+
         let cyclic_check_result =
             bytecode_verifier::CyclicModuleDependencyChecker::verify_module(module, |module_id| {
-                modules
-                    .get(module_id)
-                    .expect("Missing dependencies in cyclic checker is unexpected")
-                    .immediate_module_dependencies()
+                code_cache
+                    .get_module(module_id)
+                    .map_err(|_| PartialVMError::new(StatusCode::MISSING_DEPENDENCY))
             });
-        if cyclic_check_result.is_err() {
-            bail!(
-                "Cyclic module dependencies are detected with module {} in the loop",
-                module.self_id()
-            )
+        if let Err(cyclic_check_error) = cyclic_check_result {
+            if cyclic_check_error.major_status() == StatusCode::CYCLIC_MODULE_DEPENDENCY {
+                bail!(
+                    "Cyclic module dependencies are detected with module {} in the loop",
+                    module.self_id()
+                )
+            } else {
+                bail!(
+                    "Unexpected error happened in the cyclic module dependency checker: {}",
```

Comment:
This should probably be an `assert` rather than ` bail`, since it's an internal invariant violation.